### DISCUSSION TITLE
staging/unstaging files is extremly slow - even moving cursor in the "Unstaged changed" area stalls from time to time

### DISCUSSION THREAD

**Discussion by rbarzic** [magit/magit#4700] [Support] [answered]
Hello,

I'm using emacs+magit  on big repositories with sometimes big files (>10MB) so I'm already aware of magit issues with that.
But I don't understand why staging/unstaging file operations are so slow:

This is the result of staging one file:

Refreshing buffer ‘magit: aldebaran’...
  magit-insert-error-header                          3.507e-06 
  magit-insert-diff-filter-header                    0.00378916 
  magit-insert-head-branch-header                    0.006384917 
  magit-insert-upstream-branch-header                0.006332761 
  magit-insert-push-branch-header                    6.3609e-05 
  magit-insert-tags-header                           0.009230071 
  magit-insert-status-headers                        0.029439015 !
  magit-insert-merge-log                             0.00143017 
  magit-insert-rebase-sequence                       0.000173753 
  magit-insert-am-sequence                           8.6982e-05 
  magit-insert-sequencer-sequence                    0.000254584 
  magit-insert-bisect-output                         8.6421e-05 
  magit-insert-bisect-rest                           1.8053e-05 
  magit-insert-bisect-log                            1.8074e-05 
  magit-insert-untracked-files                       0.476792434 !!
  **_magit-insert-unstaged-changes                      36.832704611 !!_**
  magit-insert-staged-changes                        3.077116382 !!
  magit-insert-stashes                               0.001721613 
  magit-insert-unpushed-to-pushremote                8.0039e-05 
  magit-insert-unpushed-to-upstream-or-recent        0.059306684 !!
  magit-insert-unpulled-from-pushremote              1.643e-06 
  magit-insert-unpulled-from-upstream                0.003905235 
Refreshing buffer ‘magit: aldebaran’...done (40.710s)
Running magit-post-refresh-hook...done (0.000s)
Refreshing magit...done (40.712s, cached 59/91 (65%))

36 sec for magit-insert-unstaged-changes - Same for unstaging a single file.

I've tried to apply the recommendations from https://magit.vc/manual/magit/Performance.html but it does not help and I can't disable 
magit-insert-unstaged-changes either

And just moving the cursor up and down (using the keyboard) in the list of files in the "Unstaged changes" area  freezes emacs sometimes for 1 sec.

Any clue ? 

To be sure I got a clean setup, I ran with emacs using https://github.com/alphapapa/with-emacs.sh:
./with-emacs.sh --dir emacs-magit --install magit

Magit 20220527.1043 [>= 3.3.0-git], Git 2.25.1, Emacs 28.1, gnu/linux

**Comment by tarsius** [comment, marked as answer]
> big files (>10MB)
> Any clue ?

There really isn't much I can say that goes beyond [this FAQ](https://magit.vc/manual/magit/I-changed-several-thousand-files-at-once-and-now-Magit-is-unusable.html).

`(setq magit-diff-refine-hunk nil)` might help a bit.

Movement might also be slow due to Emacs' general difficulty with long lines.

**Reply by rbarzic** [reply]
What I experienced is really cursor movement slowing down when the cursor is over big files. I know it looks weird but I really checked that behavior several times. 
I have 20 files listed in the "Unstaged changes" section .  8 of those files are really big (and grouped together in the list of files). When the cursor enters this group, cursor movement is clearly slowed - It looks like magit is calling some kind of git diff command when the cursor reaches a new line - I didn't add time to debug that

**Reply by tarsius** [reply]
> It looks like magit is calling some kind of git diff command when the cursor reaches a new line - I didn't add time to debug that

Yes, and `(setq magit-diff-refine-hunk nil)` should make that stop.

**Reply by tarsius** [reply]
There is nothing I can do for you at this time.

**Reply by rbarzic** [reply]
> (setq magit-diff-refine-hunk nil) should make that stop.

Hmmm - I don't see that changing anything

Result of M-x describe-variable:

> magit-diff-refine-hunk is a variable defined in ‘magit-diff.el’.
> 
> Its value is nil

And I have still the slow cursor movement issue

**Reply by tarsius** [reply]
I am very sorry things are slow but there is nothing I can do about it.

**Reply by phil-s** [reply]
If those huge files are effectively binary files then refer to https://www.git-scm.com/docs/gitattributes#_marking_files_as_binary to prevent the diffs from being generated at all.
